src/java/org/apache/cassandra/config/registry/ConfigurationRegistry.java

```diff
@@ -0,0 +1,401 @@
+        {
+            if (initialized)
+                return;
+            properties = ImmutableMap.copyOf(defaultLoader()
+                                             .flatten(Config.class)
+                                             .entrySet()
+                                             .stream()
+                                             .map(e -> new AbstractMap.SimpleEntry<>(e.getKey(), new PropertyAdapter(configProvider, e.getValue())))
+                                             .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+            Set<String> leftConstraints = new HashSet<>(constraints.keySet());
+            leftConstraints.removeAll(properties.keySet());
+            if (!leftConstraints.isEmpty())
+                throw new ConfigurationException("Constraints are defined for non-existing properties:" + leftConstraints);
+            Set<String> leftListeners = propertyChangeListeners.values().stream()
+                                                               .map(l -> l.wrappers.keySet())
+                                                               .flatMap(Collection::stream)
+                                                               .collect(Collectors.toSet());
+            leftListeners.removeAll(properties.keySet());
+            if (!leftListeners.isEmpty())
+                throw new ConfigurationException("Listeners are defined for non-existing properties:" + leftListeners);
+            initialized = true;
+        }
+        finally
+        {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    @Override public void set(String name, @Nullable Object value)
+    {
+        lazyInit();
+        PropertyAdapter property = properties.get(name);
+        validatePropertyExists(property, name);
+        setInternal(property, value);
+    }
+
+    /**
+     * Setter for the property with the given name. Can accept {@code null} value.
+     * @param property The property.
+     * @param value The value to set.
+     */
+    private void setInternal(PropertyAdapter property, @Nullable Object value)
+    {
+        rwLock.writeLock().lock();
+        try
+        {
+            Class<?> originalType = property.getType();
+            Class<?> sourceType = value == null ? null : value.getClass();
+            Object convertedValue = value;
+            // Do conversion if the value is not null and the type is not the same as the property type.
+            if (sourceType != null && !primitiveToWrapper(originalType).equals(sourceType))
+            {
+                StringConverters converter;
+                if (sourceType.equals(String.class) && (converter = StringConverters.fromType(originalType)) != null)
+                    convertedValue = converter.fromString((String) value, originalType);
+                else
+                    throw new IllegalArgumentException(String.format("No converter found for type '%s'", originalType.getName()));
+            }
```

Comment:
this section isn't needed.  Why are we trying to support saving the wrong type?  This adds a lot of complexity and possible bugs, we should keep this super simple.

```
Object oldValue = property.getValue();
             propertyChangeListeners.get(ConfigurationListener.ChangeType.BEFORE).fireTypeCast(property.getName(), oldValue, convertedValue);
             property.setValue(convertedValue);
             propertyChangeListeners.get(ConfigurationListener.ChangeType.AFTER).fireTypeCast(property.getName(), oldValue, convertedValue);
             // This potentially may expose the values that are not safe to see in logs on production.
             logger.info("Property '{}' updated from '{}' to '{}'.", property.getName(), oldValue, convertedValue);
```